### PR TITLE
feat: hashing linkdexer

### PR DESCRIPTION
**Pull request by alanshaw**
Uses the `HashingLinkIndexer` if requested with `hash=true` querystring on the `/cid/:cid` endpoint.

e.g. https://pr13.linkdex.dag.haus/cid/bafybeibar5r2xi44meb3jyhbfs6ebllqya3o6wca4a7mbszdpfhagw22a4?hash=true

Upgrades to Node.js 18 runtime and new sst and new all the other deps.

### CODE REVIEW

File: sst.config.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+export default {
+  config (_input) {
+    return {
+      name: 'linkdex-api',
+      region: 'us-east-2'
+    }
+  },
+  stacks (app) {
+    app.setDefaultFunctionProps({ runtime: "nodejs18.x" })
```
Comment (vasco-santos) [inline sst.config.ts]:
```suggestion
    app.setDefaultFunctionProps({
      runtime: "nodejs18.x",
      nodejs: {
        format: 'esm',
        sourcemap: true
      }
    })
```

Comment (vasco-santos) [reply sst.config.ts]: https://docs.sst.dev/constructs/Function#nodejs

so that we get nice monitoring

**Review by vasco-santos — APPROVED** [review]
This looks good to me.

We should enable ESM and source maps. Noticed this was not wired up with sentry and therefore it is not super important. But I would like to see this in sentry too, can we get an issue to get it?

File: sst.config.ts (hunk 1 above)
Comment (alanshaw) [reply sst.config.ts]: ESM is the default, but I have enabled sourcemaps.

**Comment by alanshaw** [comment]
https://github.com/web3-storage/linkdex-api/issues/14

**Review by vasco-santos — APPROVED** [review]

File: services/functions/linkdex-cid.ts [hunk 2]
```diff
@@ -16,13 +17,19 @@ export interface Report extends LDReport {
...
+
 const nullReport: Report = { cars: [], structure: 'Unknown', blocksIndexed: 0, uniqueCids: 0, undecodeable: 0 }
 
 export const handler: APIGatewayProxyHandlerV2 = async (event) => {
   const cid = event.rawPath.split('/cid/')[1]
+  const hash = event.rawQueryString.includes('hash=true')
   const bucket = process.env.BUCKET_NAME ?? ''
   const s3 = new S3Client({})
-  const reporters = [new CompleteReporter(bucket, s3), new RawReporter(bucket, s3)]
+  const reporters = [new CompleteReporter(bucket, s3, { hash }), new RawReporter(bucket, s3, { hash })]
```
Comment (olizilla) [inline services/functions/linkdex-cid.ts:32]: minor but maybe a little nicer to do the query wrangling once and pass in the chosen indexer to a reporter.
```suggestion
  const index = this.options.hash ? new HashingLinkIndexer() : new LinkIndexer()
  const reporters = [new CompleteReporter(bucket, s3, index), new RawReporter(bucket, s3, index)]
```

File: sst.config.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { SSTConfig } from 'sst'
+import { Tags } from 'aws-cdk-lib'
+import { LinkdexStack } from './stacks/LinkdexStack'
+
+export default {
+  config (_input) {
+    return {
+      name: 'linkdex-api',
+      region: 'us-east-2'
```
Comment (olizilla) [inline sst.config.ts:9]: worth double checking that the mechanism to override this default by an env var hasn't changed in new sst. It'd be suprising, but we don't want to accidentally publish prod to the dev region. it's set in seed.run env config currently

**Review by olizilla — APPROVED** [review]
Good. Non-blocking comments inline

File: services/functions/linkdex-cid.ts (hunk 2 above)
Comment (alanshaw) [reply services/functions/linkdex-cid.ts:32]: The `RawReporter` needs to create an indexer for each group of CARs it considers so we cannot pass here.

**Pull request merged**